### PR TITLE
Porting desktop changes to core

### PR DESCRIPTION
**Pull request by Maoni0**
this includes the following -

+alloc lock split into SOH and LOH
+provisional mode to fix too many gen2 GCs triggered in low mem situation when the heap has heavy pinning fragmentation
+better free list usage
+premature OOM fixes
+3 new configs: GCHeapAffinitizeMask, GCHighMemPercent, GCLOHThreshold (will be documented)

I still have one perf todo for the YieldProcessor scaling factor that I have to test as the desktop and coreclr implementations are different. I'll get to it tomorrow.

### CODE REVIEW

**Comment by Maoni0** [comment]
@dotnet_bot test Windows_NT Checked longgc
@dotnet_bot test Windows_NT Release longgc
@dotnet_bot test Windows_NT Release standalone_gc
@dotnet_bot test Windows_NT Checked standalone_gc
@dotnet_bot test OSX10.12 Checked standalone_gc
@dotnet_bot test Ubuntu Checked standalone_g
@dotnet_bot test Windows_NT Debug gcsimulator
@dotnet_bot test Windows_NT Release gcsimulator
@dotnet_bot test Ubuntu Release gcsimulator
@dotnet_bot test OSX10.12 Release gcsimulator

File: src/vm/gcenv.ee.cpp [hunk 1]
```diff
@@ -1082,17 +1081,58 @@ bool GCToEEInterface::GetIntConfigValue(const char* key, int64_t* value)
       GC_NOTRIGGER;
     } CONTRACTL_END;
 
+    if (strcmp(key, "GCSegmentSize") == 0)
```
Comment (jkotas) [inline src/vm/gcenv.ee.cpp]: Do we really need to special case these 3 config settings for CoreCLR here?

File: src/gc/gc.cpp [hunk 2]
```diff
@@ -5291,45 +5334,59 @@ void set_thread_affinity_mask_for_heap(int heap_number, GCThreadAffinity* affini
...
 
 bool gc_heap::create_gc_thread ()
 {
     dprintf (3, ("Creating gc thread\n"));
+
+    GCThreadAffinity affinity;
+    affinity.Group = GCThreadAffinity::None;
+    affinity.Processor = GCThreadAffinity::None;
+
+#if !defined(FEATURE_REDHAWK) && !defined(FEATURE_CORECLR)
```
Comment (jkotas) [inline src/gc/gc.cpp]: This should not be undef ifdef.

Also, note that there is same piece of core in `gc_thread_stub`. It does not sound right to have it in both places.

File: src/gc/gc.cpp [hunk 3]
```diff
@@ -1,27 +1,50 @@
-// Licensed to the .NET Foundation under one or more agreements.
-// The .NET Foundation licenses this file to you under the MIT license.
-// See the LICENSE file in the project root for more information.
+//
```
Comment (jkotas) [inline src/gc/gc.cpp]: Undo the license change.

File: src/gc/gc.cpp [hunk 4]
```diff
@@ -1,27 +1,50 @@
...
+// 
 // #Overview
-//
-// GC automatically manages memory allocated by managed code.
-// The design doc for GC can be found at Documentation/botr/garbage-collection.md
-//
+// 
+// GC automatically manages memory allocated by managed code. 
+// The design doc for GC can be found at 
+// file:../../doc/BookOfTheRuntime/GC/GCDesign.doc
```
Comment (jkotas) [inline src/gc/gc.cpp]: Undo this change

File: src/gc/gc.cpp [hunk 5]
```diff
@@ -486,7 +533,7 @@ void log_va_msg(const char *fmt, va_list args)
     int pid_len = sprintf_s (&pBuffer[buffer_start], BUFFERSIZE - buffer_start, "[%5d]", (uint32_t)GCToOSInterface::GetCurrentThreadIdForLogging());
     buffer_start += pid_len;
     memset(&pBuffer[buffer_start], '-', BUFFERSIZE - buffer_start);
-    int msg_len = _vsnprintf_s(&pBuffer[buffer_start], BUFFERSIZE - buffer_start, _TRUNCATE, fmt, args );
+    int msg_len = _vsnprintf(&pBuffer[buffer_start], BUFFERSIZE - buffer_start, fmt, args );
```
Comment (janvorli) [inline src/gc/gc.cpp]: Could you please revert this change? I've changed that last year when I was removing all usages of unsafe printf functions from the runtime and GC.

File: src/gc/gc.cpp [hunk 6]
```diff
@@ -3015,12 +3085,13 @@ void gc_heap::fire_pevents()
     settings.record (&gc_data_global);
     gc_data_global.print();
 
-    FIRE_EVENT(GCGlobalHeapHistory_V2, gc_data_global.final_youngest_desired, 
+    FireEtwGCGlobalHeapHistory_V2(gc_data_global.final_youngest_desired,
```
Comment (janvorli) [inline src/gc/gc.cpp]: What is the reason for changing from FIRE_EVENT(GCGlobalHeapHistory_V2 to FireEtwGCGlobalHeapHistory_V2 ?

File: src/gc/gc.cpp (hunk 2 above)
Comment (Maoni0) [reply src/gc/gc.cpp]: oh.....that's totally a porting error. thanks for spotting it!

File: src/vm/gcenv.ee.cpp (hunk 1 above)
Comment (Maoni0) [reply src/vm/gcenv.ee.cpp]: the idea was we didn't want to do the rest at all if it's one of the ones the VM already did the work to parse and has the info for.

File: src/gc/gc.cpp (hunk 3 above)
Comment (Maoni0) [reply src/gc/gc.cpp]: will fix.

File: src/gc/gc.cpp (hunk 4 above)
Comment (Maoni0) [reply src/gc/gc.cpp]: will fix.

File: src/gc/gc.cpp [hunk 7]
```diff
@@ -25726,7 +26050,7 @@ void gc_heap::background_mark_phase ()
 #endif //MULTIPLE_HEAPS
         }
 
-        disable_preemptive (true);
+        disable_preemptive (TRUE);
```
Comment (janvorli) [inline src/gc/gc.cpp]: the disable_preemptive has a "bool" parameter type, so this looks like an unintended change (the same few lines below)

File: src/gc/gc.cpp [hunk 8]
```diff
@@ -27700,7 +28017,8 @@ void gc_heap::mark_through_cards_for_segments (card_fn fn, BOOL relocating)
     size_t        n_eph             = 0;
     size_t        n_gen             = 0;
     size_t        n_card_set        = 0;
-    uint8_t*      nhigh             = (relocating ? heap_segment_plan_allocated (ephemeral_heap_segment) : high);
+    uint8_t*         nhigh             = (relocating ?
```
Comment (janvorli) [inline src/gc/gc.cpp]: A nit - alignment

File: src/gc/gc.cpp [hunk 9]
```diff
@@ -33443,17 +33809,58 @@ HRESULT GCHeap::Initialize ()
...
-    nhp = min (nhp, MAX_SUPPORTED_CPUS);
+                if (gc_thread_affinity_mask)
+                {
+                    pmask &= gc_thread_affinity_mask;
+                }
+
+                process_mask = pmask;
+
+                unsigned int set_bits_in_pmask = 0;
+                while (pmask)
```
Comment (janvorli) [inline src/gc/gc.cpp]: A nit - an alternative if-less way to get number of set bits would be
```c++
while (pmask != 0)
{
    set_bits_in_pmask++;
    pmask &= (pmask - 1);
}
```

File: src/gc/gcpriv.h [hunk 10]
```diff
@@ -1,6 +1,10 @@
-// Licensed to the .NET Foundation under one or more agreements.
-// The .NET Foundation licenses this file to you under the MIT license.
-// See the LICENSE file in the project root for more information.
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (janvorli) [inline src/gc/gcpriv.h]: The licence header should stay as it was.

File: src/gc/gcpriv.h [hunk 11]
```diff
@@ -3062,11 +3105,13 @@ class gc_heap
...
 
+#if defined (_DEBUG) && defined (VERIFY_HEAP)
     PER_HEAP
-    BOOL        verify_pinned_queue_p;
+    BOOL       verify_pinned_queue_p;
+#endif // _DEBUG && VERIFY_HEAP
 
     PER_HEAP
-    uint8_t*    oldest_pinned_plug;
+    uint8_t*       oldest_pinned_plug;
```
Comment (janvorli) [inline src/gc/gcpriv.h]: Looks like unintentional change.

File: src/gc/gc.cpp (hunk 5 above)
Comment (Maoni0) [reply src/gc/gc.cpp]: porting error, thanks for spotting!

File: src/gc/gc.cpp (hunk 6 above)
Comment (Maoni0) [reply src/gc/gc.cpp]: all FireEtw* were changed to FIRE_EVENT for LocalGC because we wanna check if it's enabled on the GC side.

File: src/gc/gc.cpp (hunk 7 above)
Comment (Maoni0) [reply src/gc/gc.cpp]: will fix.

File: src/gc/gc.cpp (hunk 8 above)
Comment (Maoni0) [reply src/gc/gc.cpp]: will fix

File: src/gc/gc.cpp (hunk 9 above)
Comment (Maoni0) [reply src/gc/gc.cpp]: yep that's more efficient - but it makes no actual difference cause other calls will dominate so I'll leave it the same as on desktop (code is a bit easier to read).

File: src/gc/gcpriv.h (hunk 11 above)
Comment (Maoni0) [reply src/gc/gcpriv.h]: will fix

**Comment by Maoni0** [comment]
this addresses all feedback above + a bunch of things I found while going through the changes + correct SetYieldProcessorScalingFactor for core.

**Comment by Maoni0** [comment]
@dotnet_bot test Windows_NT Checked longgc
@dotnet_bot test Windows_NT Release longgc
@dotnet_bot test Windows_NT Release standalone_gc
@dotnet_bot test Windows_NT Checked standalone_gc
@dotnet_bot test OSX10.12 Checked standalone_gc
@dotnet_bot test Ubuntu Checked standalone_g
@dotnet_bot test Windows_NT Debug gcsimulator
@dotnet_bot test Windows_NT Release gcsimulator
@dotnet_bot test Ubuntu Release gcsimulator
@dotnet_bot test OSX10.12 Release gcsimulator

**Comment by jkotas** [comment]
```
src/gc/gc.cpp:12880:9: error: cannot jump from this goto statement to its label
        goto exit;
        ^
src/gc/gc.cpp:12885:17: note: jump bypasses variable initialization
    GCSpinLock* msl = loh_p ? &more_space_lock_loh : &more_space_lock_soh;
                ^
In file included from src/gc/gcwks.cpp:26
```

**Comment by Maoni0** [comment]
thanks Jan! msvc compiler does not complain this. ahh well. fixed and resubmitted.

**Comment by Maoni0** [comment]
thanks Jan! msvc compiler does not complain about this. ahh well. fixed and resubmitted.

**Comment by Maoni0** [comment]
@dotnet-bot test Ubuntu x64 Checked Innerloop Build and Test

**Comment by Maoni0** [comment]
@dotnet-bot test CentOS7.1 x64 Debug Innerloop Build

**Comment by Maoni0** [comment]
@dotnet-bot test Windows_NT x86 Checked Innerloop Build and Test (Jit - TieredCompilation=0)

**Comment by Maoni0** [comment]
@dotnet_bot test Windows_NT Debug gcsimulator
@dotnet_bot test Windows_NT Release gcsimulator

**Comment by Maoni0** [comment]
@dotnet-bot test test Ubuntu Release longgc

**Comment by Maoni0** [comment]
@dotnet-bot test Ubuntu Release longgc

**Comment by Maoni0** [comment]
@dotnet_bot test Windows_NT x64 Debug gcsimulator

**Comment by Maoni0** [comment]
@dotnet-bot test Windows_NT x86 Checked Innerloop Build and Test (Jit - TieredCompilation=0)

**Comment by mmitche** [comment]
@dotnet-bot test this please

**Comment by mmitche** [comment]
@dotnet-bot test Windows_NT x86 Checked Innerloop Build and Test (Jit - TieredCompilation=0)

**Comment by Maoni0** [comment]
@dotnet_bot test Windows_NT Release longgc

**Comment by mmitche** [comment]
@dotnet-bot test Windows_NT x86 Checked Innerloop Build and Test

**Comment by mmitche** [comment]
@dotnet-bot test Windows_NT x64 Release longgc

**Comment by mmitche** [comment]
@dotnet_bot test Windows_NT x64 Checked longgc
@dotnet_bot test Windows_NT x64 Release longgc
@dotnet_bot test Windows_NT x64 Release standalone_gc
@dotnet_bot test Windows_NT x64 Checked standalone_gc
@dotnet_bot test OSX10.12 x64 Checked standalone_gc
@dotnet_bot test Ubuntu x64 Checked standalone_g
@dotnet_bot test Windows_NT x64 Debug gcsimulator
@dotnet_bot test Windows_NT x64 Release gcsimulator
@dotnet_bot test Ubuntu x64 Release gcsimulator
@dotnet_bot test OSX10.12 x64 Release gcsimulator

**Comment by mmitche** [comment]
@dotnet_bot test Windows_NT x64 Checked longgc
@dotnet_bot test Windows_NT x64 Release longgc
@dotnet_bot test Windows_NT x64 Release standalone_gc
@dotnet_bot test Windows_NT x64 Checked standalone_gc
@dotnet_bot test OSX10.12 x64 Checked standalone_gc
@dotnet_bot test Ubuntu x64 Checked standalone_g
@dotnet_bot test Windows_NT x64 Debug gcsimulator
@dotnet_bot test Windows_NT x64 Release gcsimulator
@dotnet_bot test Ubuntu x64 Release gcsimulator
@dotnet_bot test OSX10.12 x64 Release gcsimulator

**Comment by Maoni0** [comment]
@dotnet_bot test Windows_NT x64 Checked longgc
@dotnet_bot test Windows_NT x64 Release longgc
@dotnet_bot test Windows_NT x64 Release standalone_gc
@dotnet_bot test Windows_NT x64 Checked standalone_gc
@dotnet_bot test OSX10.12 x64 Checked standalone_gc
@dotnet_bot test Ubuntu x64 Checked standalone_g
@dotnet_bot test Windows_NT x64 Debug gcsimulator
@dotnet_bot test Windows_NT x64 Release gcsimulator
@dotnet_bot test Ubuntu x64 Release gcsimulator
@dotnet_bot test OSX10.12 x64 Release gcsimulator
